### DISCUSSION TITLE
Regridding 3D fields (e.g., GLORYS --> ROMS)

### DISCUSSION THREAD

**Discussion by NoraLoose** [pangeo-data/xESMF#389] [Q&A] [answered]
I would like to regrid 3D fields from a regular lat / lon / z-level grid (e.g., GLORYS) to a curvilinear grid with terrain-following "s"-coordinates (e.g., ROMS). Is anyone aware of a regridding tool that support this kind of regridding?

I am aware of [`xarray.DataArray.interp`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.interp.html) which supports 3D interpolation, but only for simple methods (like `linear` or `nearest`). Another problem is that the xarray method is computationally not very efficient if used for many time steps because it does not reuse pre-computed weights as `xesmf` for the regridding. 

I'm interested to hear what other folks do!

**Comment by raphaeldussin** [comment, marked as answer]
Hi @NoraLoose 

I don't know if there is an all-in-one code to do that but this is what I've done or seen in the past (e.g. https://github.com/ESMG/pyroms/tree/python3/pyroms_toolbox/pyroms_toolbox/CGrid_GLORYS)

* work level by level
* Extrapolate glorys to fill in land values (https://github.com/ajdawson/gridfill, https://github.com/raphaeldussin/HCtFlood)
* use xESMF to interpolate onto the ROMS horizontal grid
* get ROMS (3d) layer depths using the ROMS Vtransform/Vstretching equations
* interpolate from z level to the ROMS layer depths (xgcm may support this or https://github.com/raphaeldussin/VCR for a piecewise constant conservative interpolation)

**Reply by NoraLoose** [reply]
Thanks for your quick response @raphaeldussin! Follow-up question: Do you follow the same approach for creating the ROMS boundary forcing? I can think of two approaches:
1. Use your steps to do a 3D interpolation, and then just select the 4 boundaries
2. Interpolate onto the 4 2D (one horizontal, one vertical D) boundaries directly never using `xesmf`.

The context of all this is that we are building a new tool box to create ROMS input forcing here: https://github.com/CWorthy-ocean/roms-tools

**Reply by raphaeldussin** [reply]
for OBC, I used the `locstream` option of xesmf and then you just need to fill in the land using ffill or bfill. This will go much faster than doing the 3d interp just for the few segments you need. I wrote this example for MOM6 but the same idea for ROMS + repeat the vertical interpolation. Example: https://github.com/ESMG/regionalMOM6_notebooks/blob/master/creating_obc_input_files/CCS1_OBC_from_global_MOM6.ipynb

**Reply by raphaeldussin** [reply]
Also throwing my 2 cents: all the regional domains have their own unique features so I never saw a successful "one size fits all" solution. I advocate for publishing notebooks describing the algorithm so that anyone trying to generate a new domain can benefit from the various ideas used before.

**Reply by NoraLoose** [reply]
Cool, thanks so much @raphaeldussin for your helpful answers!! 🚀
